### PR TITLE
Add database support

### PR DESCRIPTION
**Pull request by j-waters**
I was going to start work on #81 but decided to get some basic database support working first. All state is now held in an SQLite database. I've also made it a lot easier to add and remove coins on the fly. 

I've tested this on several full trades and it's working fine, but it could probably do with some more thorough testing. Also, I updated some of the config reading code to work with the new system but didn't look too hard at how it worked, so that could probably be simplified now.

### CODE REVIEW

**Comment by j-waters** [comment]
The schema is very simple at the moment:

* `coins`
  * `symbol` - str, private key
  * `enabled` - bool
* `current_coin_history` - the most recent entry is the current coin
  *  `coin_id` - foreign key linking to coins table
  * `datetime`
* `pairs` - all the possible pairs
  * `from_coin_id` - foreign key, the 'current' coin
  * `to_coin_id` - the coin that you could move to from the other coin
  * `ratio` - required price ratio for the trade to occur

**Comment by j-waters** [comment]
This basically adds no functionality, it's a straight port from the backup files to the database. I'd like to add a `trades` table, which would act as both a history of trades made as well as provide some crash resistance - we could keep track of the state of trades so that if the script is stopped before a trade is completed we can carry on.

And as I said in #81, we can have tables for keeping track of the current ratios between all the coins, current coin values, current market values etc that could then be used by whatever UI ends up being made

**Comment by edeng23** [comment]
Looks good, I'll start testing it right now :)

**Comment by porter707** [comment]
This is great, I've given this a little test (not made any trades with it) and the initial startup all appears to work and the Database looks good.

One request/suggestion would be an upgrade path for existing installs, so it detects an existing current_coin_table and loads that data into the pairs table. Then rename the current_coin_table file so it doesn't get picked up again on the next launch, then the user can manually remove that file if they are happy the upgrade went smoothly.

**Comment by rel-s** [comment]
I agree with @porter707, it would save us a hefty amount of new issues to deal with...

**Comment by edeng23** [comment]
@porter707 Good point, we wouldn't want existing users to lose their coin ratios.

**Comment by j-waters** [comment]
Good idea I'll do that now

**Comment by slyths** [comment]
@j-waters a bit off topic, but are you on the discord server and if you are, could you please tell me the nick you are using on discord. I would like to update your role.

**Comment by rel-s** [comment]
Also worth noting that this PR makes #77 obsolete, so maybe we don't have to merge it.

**Comment by j-waters** [comment]
Okay that should be working, I've kept the docker volume mounts so that migrating works for docker users although they will be able to be removed at some point.

As for #77 it might still be good to have one single config file instead of two

@slyths I am although I haven't looked at it very much! I'm Inazuma#4514

**Comment by j-waters** [comment]
I think this also makes https://github.com/edeng23/binance-trade-bot/pull/85 obselete

**Comment by chung1905** [comment]
Yes, going to close it now.
Since this project is getting big, I think it's important to do a proper code style.

File: README.md [hunk 1]
```diff
@@ -56,7 +56,10 @@ For more information about Telegram bots refer to [Telegram's official documenta
 
 ### Docker
 
-`docker-compose up`
+```shell
+mkdir data
+docker-compose up
+```
```
Comment (chung1905) [inline README.md]: We can just create a .gitignore in data directory.
Can save some issue in repo.

Comment (j-waters) [reply README.md]: Great point

**Comment by j-waters** [comment]
Been using this branch for a while now and haven't encountered any issues, I'd be happy for it to be merged. I've continued on from this in PRs #106 and #107

**Comment by edeng23** [comment]
I saw, I'm probably going to merge it very soon. I'm nearly done with my tests.

**Comment by edeng23** [comment]
@j-waters I found two bugs in the "migrate old state" logic. 
1. When you call session.add(pair) on line 426 of crypto_trading.py, I believe that you are trying to add a pair that is already in the database causing the following error:
_"sqlalchemy.exc.InvalidRequestError: Can't attach instance <Coin at 0x7fffeb9babe0>; another instance with key (<class 'models.coin.Coin'>, ('XLM',), None) is already present in this session."_
2.  The code crashes when trying to add coins that are referencing themselves in the old list (Coin A -> Coin A, ratio=1). I take the blame for writing ugly code in the first place. That being said, in order to make your code backwards compatible, you should bypass this when loading the old coin table.
Other than these two issues, everything else ran perfectly. Please fix these problems, I'm eagerly waiting to merge your code :)

**Comment by j-waters** [comment]
Ah thanks, that should be fixed!

**Pull request merged**